solutions/usaco-989.mdx

```diff
@@ -2,11 +2,215 @@
 id: usaco-989
 source: USACO Bronze 2020 January
 title: Race
-author: Sathvik Chundru, Kevin Sheng
+author: Sathvik Chundru, Kevin Sheng, Jonathan Paulson
 ---
 
 [Official Analysis (C++)](http://www.usaco.org/current/data/sol_race_bronze_jan20.html)
 
+## Video Solution
+
+By Jonathan Paulson
+
+<Youtube id="4KcvKIJP9KY" />
+
+<Spoiler title="Video Solution Code">
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <iostream>
+#include <vector>
+#include <cassert>
+#include <cstdio>
+using namespace std;
+using ll = long long;
+
+// S(n) = 1+2+...+n = n*(n+1)/2
+ll S(ll n) {
+  return n*(n+1)/2;
+}
+// A+(A+1)+(A+2)+...+B = \sum_{i=A}^B i
+ll sum(ll A, ll B) {
+  return A<=B ? S(B)-S(A-1) : 0;
+  //return A<=B ? (A+B)*(B-A+1)/2 : 0;
+}
+// Given that we start at speed S, can we slow down to speed X in at most L meters?
+// How many meters does it take to stop given current speed is S and target is X?
+// (S-1) + (S-2) + ... + X meters
+// We are allowed to go over, so as long as we have (S-1)+(S-2)+...+(X+1)+1 m we are OK.
+bool ok(ll S, ll X, ll L) {
+  if(S <= X) { return true; }
+  ll need = sum(X+1, S-1)+1;
+  bool ans = (need <= L);
+  return ans;
+}
+ll solve(ll k, ll x) {
+  // Greedy. Go as fast as you can but make sure you can stop in time.
+  ll speed = 0;
+  ll left = k;
+  ll t = 0;
+  while(left>0) {
+    if(ok(speed+1, x, left-(speed+1))) {
+      // speed up if we can
+      speed += 1;
+    } else if(ok(speed, x, left-speed)) {
+      // otherwise keep the same speed if we can
+    } else {
+      // slow down if we have to
+      // check that we have enough space left to stop in time.
+      assert(ok(speed-1, x, left-(speed-1)));
```

Comment:
tabs

Reply:
oh bruh the auto-formatter didn't catch that